src/main.rs

```diff
@@ -120,24 +122,44 @@ async fn main() -> Result<()> {
     tokio::spawn(health_status(health_reporter, Arc::clone(&pool)));
 
     // start replication
-    tokio::spawn(replicate(replication_state));
-    // start unknown reaper - removes replication objects for public_keys that moved from Unkown -> Local
-    tokio::spawn(reap_unknown_keys(Arc::clone(&pool), Arc::clone(&cache)));
+    if config.replication_enabled {
+        tokio::spawn(replicate(replication_state));
+        // start unknown reaper - removes replication objects for public_keys that moved from Unknown -> Local
+        tokio::spawn(reap_unknown_keys(Arc::clone(&pool), Arc::clone(&cache)));
+    }
+
     // start datadog reporter
     let (datadog_sender, datadog_receiver) = channel::<MinitraceSpans>(10);
-    tokio::spawn(report_datadog_traces(datadog_receiver, config.dd_agent_host.clone(), config.dd_agent_port, config.dd_service_name.clone()));
+    if let Some(ref dd_config) = config.dd_config {
+        tokio::spawn(report_datadog_traces(
+            datadog_receiver,
+            dd_config.agent_host,
+            dd_config.agent_port,
+            dd_config.service_name.clone(),
+        ));
+    }
 
     log::info!("Starting server on {:?}", &config.url);
 
     // TODO add server fields that make sense
-    Server::builder()
-        .layer(MinitraceGrpcMiddlewareLayer::new(datadog_sender))
-        .add_service(health_service)
-        .add_service(PublicKeyServiceServer::new(public_key_service))
-        .add_service(MailboxServiceServer::new(mailbox_service))
-        .add_service(ObjectServiceServer::new(object_service))
-        .serve(config.url)
-        .await?;
+    if config.dd_config.is_some() {
+        Server::builder()
+            .layer(MinitraceGrpcMiddlewareLayer::new(datadog_sender))
+            .add_service(health_service)
+            .add_service(PublicKeyServiceServer::new(public_key_service))
+            .add_service(MailboxServiceServer::new(mailbox_service))
+            .add_service(ObjectServiceServer::new(object_service))
+            .serve(config.url)
+            .await?;
+    } else {
+        Server::builder()
+            .add_service(health_service)
+            .add_service(PublicKeyServiceServer::new(public_key_service))
+            .add_service(MailboxServiceServer::new(mailbox_service))
+            .add_service(ObjectServiceServer::new(object_service))
+            .serve(config.url)
+            .await?;
+    }
```

Comment:
Minor nit: possibly factor out common operations since there's a lot of repetition

```
 Server::builder()
    .add_service(health_service)
    .add_service(PublicKeyServiceServer::new(public_key_service))
    .add_service(MailboxServiceServer::new(mailbox_service))
    .add_service(ObjectServiceServer::new(object_service))
   ...
```

Reply:
I attempted a few different ways and couldn't get the types to check. I tried `Box`ing and checked for a common `impl` trait that could be leveraged with no luck.

```rust
Server::builder()
```
and

```rust
Server::builder().layer(MinitraceGrpcMiddlewareLayer::new(datadog_sender))
```

Reply:
@scirner22 That's Rust I guess